### PR TITLE
Check root return in ReturnTypeFromReturnNewRector

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+final class SkipPossibleVoidWithEarlyReturn
+{
+    public function bar()
+    {
+        if (random_int(0,1)) {
+            return;
+        }
+
+        return new self();
```
Comment (samsonasik) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: I think `ReturnTypeInferer` service is more mature way to detect that, also allow verify code inside `while (true)`, `do { } while`, `if else return`, deep condition etc.

Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: What do you suggest?

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: The existing code:

```php
$returnType = $this->returnTypeInferer->inferFunctionLike($node);
if ($returnType instanceof UnionType) {
   return null;
}
```

as current verify `UnionType` is already fine for me.

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: see existing code https://getrector.com/demo/e688fb7f-02ce-4e8b-8015-73877dd62a9a already skipped as expected.

Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: That's missuse of unrelated service that accidentally works here :)

This rule is only to add types for instance of `new` when possible. Union suggests PHP 8 is involved, which is not valid here.

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: That means `SilentVoidResolver` should be used instead, that's more correct service to be used

Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: I think that's not neccesary either, as it checks inverted condition.
Ideally let me know failing fixture to cover, I can figure rest :)

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/skip_possible_void_with_early_return.php.inc:15]: ok 👍

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/if_new_else_new.php.inc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+final class IfNewElseNew
+{
+    public function action()
+    {
+        if (mt_rand(0, 1)) {
+            return new FirstResponse();
+        }
+
+        return new SecondResponse();
```
Comment (samsonasik) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/if_new_else_new.php.inc:18]: new fixture for `else` syntax seems needed:

```php
        if (mt_rand(0, 1)) {
            return new FirstResponse();
        } else {
            return new SecondResponse();
       }
```

`SilentVoidResolver->hasSilentVoid($node)` should can detect that, if you want to consider...

Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/if_new_else_new.php.inc:18]: I'll look into it in next iteration

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnNewRector/Fixture/if_new_else_new.php.inc:18]: I already create new PR for that:

- https://github.com/rectorphp/rector-src/pull/6104

**Pull request merged**